internal/cloud/aws/logger.go

```diff
@@ -6,20 +6,192 @@ SPDX-License-Identifier: AGPL-3.0-only
 
 package aws
 
-// TODO: Implement for AWS.
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
+	logs "github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs"
+	"github.com/aws/aws-sdk-go-v2/service/cloudwatchlogs/types"
+	"k8s.io/utils/clock"
+)
 
 // Logger is a Cloud Logger for AWS.
-type Logger struct{}
+// Log messages are collected and periodically flushed to AWS Cloudwatch Logs.
+type Logger struct {
+	api logAPI
+
+	groupName  string
+	streamName string
+
+	logs          []types.InputLogEvent
+	sequenceToken *string
+
+	flushMux sync.Mutex
+	interval time.Duration
+	clock    clock.WithTicker
+	wg       *sync.WaitGroup
+	stopCh   chan struct{}
+}
 
 // NewLogger creates a new Cloud Logger for AWS.
-func NewLogger() *Logger {
-	return &Logger{}
+func NewLogger() (*Logger, error) {
+	cfg, err := config.LoadDefaultConfig(context.Background(), config.WithEC2IMDSRegion())
+	if err != nil {
+		return nil, err
+	}
+	client := logs.NewFromConfig(cfg)
+
+	l := &Logger{
+		api:      client,
+		interval: time.Second,
+		clock:    clock.RealClock{},
+		wg:       &sync.WaitGroup{},
+		stopCh:   make(chan struct{}, 1),
+	}
+
+	if err := l.createStream(imds.New(imds.Options{})); err != nil {
+		return nil, err
+	}
+
+	l.wg.Add(1)
+	go l.flushLoop()
+
+	return l, nil
 }
 
-// Disclose is not implemented for AWS.
-func (l *Logger) Disclose(msg string) {}
+// Disclose adds a message to the log queue.
+// The messages are flushed periodically to AWS Cloudwatch Logs.
+func (l *Logger) Disclose(msg string) {
+	l.logs = append(l.logs, types.InputLogEvent{
+		Message:   aws.String(msg),
+		Timestamp: aws.Int64(time.Now().UnixMilli()),
+	})
+}
 
-// Close is a no-op.
+// Close flushes the logs a final time and stops the flush loop.
 func (l *Logger) Close() error {
+	l.stopCh <- struct{}{}
+	l.wg.Wait()
+	return l.flushLogs()
+}
+
+// flushLogs flushes the aggregated log messages to AWS Cloudwatch Logs.
+func (l *Logger) flushLogs() error {
+	// make sure only one flush operation is running at a time
+	l.flushMux.Lock()
+	defer l.flushMux.Unlock()
+
+	if len(l.logs) == 0 {
+		return nil // no logs to flush
+	}
+
+	res, err := l.api.PutLogEvents(context.Background(), &logs.PutLogEventsInput{
+		LogEvents:     l.logs,
+		LogGroupName:  &l.groupName,
+		LogStreamName: &l.streamName,
+		SequenceToken: l.sequenceToken,
+	})
+	if err != nil {
+		// If the flush operation was called on a pre-existing stream,
+		// or another operation sent logs to the same stream,
+		// the sequence token may not be set correctly.
+		// We can retrieve the correct sequence token from the error message.
+		var sequenceErr *types.InvalidSequenceTokenException
+		if errors.As(err, &sequenceErr) {
+			l.sequenceToken = sequenceErr.ExpectedSequenceToken
+			l.flushMux.Unlock()
+			err = l.flushLogs()
+			l.flushMux.Lock()
```

Comment:
this doesn't work.

Reply:
What do you mean by "this doesn't work"?
Because unless there is some weird side effect created by this, this does what I expect it to do.
`flushLogs` is called again without causing a panic by misusing the mutex.

If the unlocking/relocking is a problem, we could move the repeated flush call to a deferred function call after the mutex has been unlocked.